silkrpc/grpc/unary_rpc.hpp

```diff
@@ -0,0 +1,115 @@
+/*
+    Copyright 2022 The Silkrpc Authors
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+        http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+*/
+
+#ifndef SILKRPC_GRPC_UNARY_RPC_HPP_
+#define SILKRPC_GRPC_UNARY_RPC_HPP_
+
+#include <memory>
+#include <system_error>
+#include <utility>
+
+#include <silkrpc/config.hpp>
+
+#include <agrpc/rpc.hpp>
+#include <asio/compose.hpp>
+#include <asio/experimental/append.hpp>
+#include <grpcpp/grpcpp.h>
+
+#include <silkrpc/grpc/dispatcher.hpp>
+#include <silkrpc/grpc/error.hpp>
+#include <silkrpc/grpc/util.hpp>
+#include <silkrpc/common/log.hpp>
+
+namespace silkrpc {
+
+namespace detail {
+struct DoneTag {
+};
+} // namespace detail
+
+template<auto Rpc>
+class UnaryRpc;
+
+template<
+    typename Stub,
+    typename Request,
+    template<typename> typename Reader,
+    typename Reply,
+    std::unique_ptr<Reader<Reply>>(Stub::*Async)(grpc::ClientContext*, const Request&, grpc::CompletionQueue*)
+>
+class UnaryRpc<Async> {
```

Comment:
If we assume that the only completion token being used is `asio::use_awaitable` then this entire class can be replaced by:

```cpp
template <class RPC, class CompletionExecutor = asio::system_executor>
auto unary_rpc(agrpc::GrpcContext& grpc_context, typename RPC::Stub& stub, const typename RPC::Request& request,
               const CompletionExecutor& completion_executor = asio::system_executor{})
    -> asio::awaitable<typename RPC::Response>
{
    grpc::ClientContext context;
    typename RPC::Response response;
    auto rpc = co_await RPC::request(grpc_context, stub, context, request, response,
                                     asio::bind_executor(completion_executor, asio::use_awaitable));
    if (!rpc.ok())
    {
        // In your code you are first turning the grpc::Status to a asio::error_code which is then thrown by Asio. I
        // think it would be easier to just throw the desired exception type and message directly, example:
        throw std::runtime_error(rpc.error_message());
    }
    co_return response;
}
```

with asio-grpc's new high-level client API.

And on the call side we would write:

```cpp
using RPC = agrpc::RPC<&::remote::ETHBACKEND::StubInterface::AsyncEtherbase>;
const auto reply = co_await silkrpc::unary_rpc<RPC>(grpc_context_, *stub, /*::remote::EtherbaseRequest*/{}, executor_);
```

Reply:
In terms of high-level API design for asio-grpc, I suppose you are planning to support any completion token. For sure this function can be a convenient all-in-one for `asio::awaitable`.

Reply:
So you would expect that everything is heap allocated so that a user can write:

```cpp
using RPC = agrpc::RPC<&remote::ETHBACKEND::StubInterface::AsyncEtherbase>;
RPC rpc{grpc_context};   // owns ClientContext and RPC::Response
rpc.client_context().set_deadline();
// callback as an example needs the lifetime extension of ClientContext and Response in the rpc
// that's why heap allocation is required
rpc.request(stub, request, [](std::expected<RPC::Response, grpc::Status> response) {
});
```

Not sure whether I like that so much. I don't want the new API to be a performance burden for users of coroutines or sender/receiver. I could allocate space for the `Response` object within the operation state of the current RPC::request call, I guess I will simply add an overload to the current API for that.

Reply:
I completely agree that new API should _not_ place a performance burden when it is not necessary, so an overload is exactly the solution I would suggest for `asio-grpc` proposal to be complete.
To be completely clear, I'm going to use the coroutine API in Silkrpc so my suggestion is just a hint to make asio-grpc as complete as possible because, well, I love asio-grpc :smile: